### ISSUE TITLE
declares globals and exports in .d.ts file not recognized

### ISSUE THREAD

**Issue by KraXen72** [quick-lint/quick-lint-js#1172]
sveltekit projects have a `app.d.ts` file. i use it like this:
```ts
declare global {
	interface NavLink {
		type: 'nav',
		title: string,
		href: string,
	}
	interface SubmenuSeparator {
		type: 'separator'
	}
}
// ...
export {};
```

However, the global types are not recognized and show:
![image](https://github.com/quick-lint/quick-lint-js/assets/21956756/e0e3c3d5-d2d7-4465-8a36-4ddea070d46e)
  
 tsconfig:  
 ```jsonc
{
	"extends": "./.svelte-kit/tsconfig.json",
	"compilerOptions": {
		"allowJs": true,
		"checkJs": true,
		"esModuleInterop": true,
		"forceConsistentCasingInFileNames": true,
		"resolveJsonModule": true,
		"skipLibCheck": true,
		"sourceMap": true,
		"strict": true,
		"target": "es2021"
	}
	// Path aliases are handled by https://kit.svelte.dev/docs/configuration#alias
	//
	// If you want to overwrite includes/excludes, make sure to copy over the relevant includes/excludes
	// from the referenced tsconfig.json - TypeScript does not merge them in
}
```
the "./.svelte-kit/tsconfig.json" is auto-generated / changes a lot, better to check a bare sveltekit starter project.

**Comment by vegerot** [comment]
I think the documentation needs to be improved.

1. https://quick-lint-js.com/errors/E0214/ should mention how to add a global identifier to the config
2. It should be easier to link to https://quick-lint-js.com/config/#_examples (I had to inspect element to get the id)

**Comment by KraXen72** [comment]
isn't globals just for declaring global variables one by one, or from enviroment sets? rather, i would need quick lint to recognize all the global types declared in app.d.ts and accept them as global

**Comment by strager** [comment]
quick-lint-js currently does not have a mechanism to take global variables declared in one file (app.d.ts in your case) and make them visible in another file.

Some possible solutions:
1. Rewrite your code to use `import` instead of globals.
2. Manually replicate app.d.ts to your quick-lint-js.config.
3. Use a tool which replicates app.d.ts's globals to your quick-lint-js.config. (This tool does not yet exist.)
4. Use a `"globals-files"` feature in quick-lint-js.config to specify source files. You would list `"globals-files": ["src/app.d.ts"]`. (This feature does not yet exist.)
5. Have quick-lint-js auto-scan for .ts and .d.ts files which declare globals, similar to what TypeScript's compiler does. (This feature does not yet exist.)

@KraXen72 Which solutions sound good to you?

**Comment by KraXen72** [comment]
while 1 is a good short term solution, i think 5 would be the best - scanning, relevant .d.ts files in the project (as far as i understand it, quicklint already does scanning and parsing in these files, since i get quicklint warnings in them), and parsing declare global blocks or other ways to declare global types, so quicklint recognizes global types and supresses the unknown type warning.

until this is implemented, a stopgap solution could be adding the option to disable the 'unknown type' warning, and leave known/unknown logic to the typescript LSP. I think many people use quicklint alongside the ts lsp, and not instead of it.

while importing and exporting types is likely a better way to manage types than globals, for a lot of projects i either don't have that many types (they fit in 1 file), or i don't want to put in the extra effort of import/export management for types too. yes, autoimport extensions exist  and they do help, but i found they often make mistakes when modifying existing imports, or just refuse to create 'import type {}' imports.

**Comment by strager** [comment]
> as far as i understand it, quicklint already does scanning and parsing in these files, since i get quicklint warnings in them

quick-lint-js only reports diagnostics for files you open in your editor. quick-lint-js does not scan all files in your project automatically.

I did want to implement project-wide analysis, but only by following `import` statements. Your use case suggests that I need to follow more than just `import` statements. I'll keep this in consideration when designing the analysis.

> a stopgap solution could be adding the option to disable the 'unknown type' warning, and leave known/unknown logic to the typescript LSP.

#1174

---

In the next few months, I don't plan on implementing a proper fix for this issue. If you want to avoid the warnings but continue using quick-lint-js, you can [configure quick-lint-js](https://quick-lint-js.com/config/#_examples).

**Comment by KraXen72** [comment]
for now, i'm using oxc linter, but i like quick-lint more - i'll try configuring it in a bit. thanks!

**Comment by SuperLlama88888** [comment]
I'd like to mention that this issue also affects many TypeScript types for the DOM/other web APIs which aren't defined in the specs.

E.g. TypeScript knows that `OffscreenRenderingContextId` exists (defined in `lib.dom.d.ts`), but quick-lint-js doesn't.
<img width="852" height="99" alt="Image" src="https://github.com/user-attachments/assets/80dbb2e0-68f5-4b7f-8380-e09c4fb6b7d8" />
